File: ros/ws02/first_ws/src/use_custom4_msg_pkg/src/subscriber_cus4_msg.cpp
//
// Created by zcb on 5/3/21.
//

#include <iostream>
#include <ros/ros.h>
#include <custom4_msgs/Team.h>


void topicCallback(const custom4_msgs::Team::ConstPtr &msg) {
    std::cout << msg->name << " || " << msg->leader.name << " " << msg->leader.age << " " << msg->intro.data
              << std::endl;
    std::cout << "members: " << std::endl;
    auto mems = msg->members;
    for (auto i = mems.begin(); i != mems.end(); i++) {
        std::cout << i->name << " " << i->age << std::endl;
    }
    std::cout << "===" << std::endl;
}

int main(int argc, char **argv) {

    std::string nodeName = "subscriber_node";
    ros::init(argc, argv, nodeName);
    ros::NodeHandle node;

    std::string topicName = "/zcb01/topic";
    ros::Subscriber sub = node.subscribe(topicName, 1000, topicCallback);

    ros::spin();
}

File: 06ros/ws02/first_ws/src/use_custom4_msg_pkg/scripts/subscriber_cus4_msg.py
#! /usr/bin/env python
# coding:utf8

import rospy
from custom4_msgs.msg import Team


def topic_callback(msg):
    if not isinstance(msg, Team):
        return
    print "%s %s %d %s" % (msg.name, msg.leader.name, msg.leader.age, msg.intro.data)
    print "members"
    for mem in msg.members:
        print "%s %d" % (mem.name, mem.age)

    print "==="


if __name__ == '__main__':
    nodeName = "subscriber_node"
    rospy.init_node(nodeName)

    topicName = "/zcb01/topic"
    rospy.Subscriber(topicName, Team, topic_callback)

    rospy.spin()
